### ISSUE TITLE
cloud-function error: Could not load the default credentials

### ISSUE THREAD

**Issue by dan-j** [google/google-auth-library-nodejs#320]
I'm currently seeing sporadic errors, in multiple Firebase Cloud Functions, with the following stack:

```
Error: Unexpected error while acquiring application default credentials: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information.
    at GoogleAuth.<anonymous> (/user_code/node_modules/@google-cloud/firestore/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:249:31)
    at step (/user_code/node_modules/@google-cloud/firestore/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:47:23)
    at Object.next (/user_code/node_modules/@google-cloud/firestore/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:28:53)
    at fulfilled (/user_code/node_modules/@google-cloud/firestore/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:19:58)
    at process._tickDomainCallback (internal/process/next_tick.js:135:7)
```

I know this shouldn't happen, because default credentials should work in a cloud function (and it does sometimes, but also sometimes produces errors).

**TL;DR;** I **_think_** I'm using 1.3.1 (see below for explanation for why I'm not sure), and if so, I don't think the problem is actually in this library, it's more likely in `gcp-metadata`. But the way this library is using `gcp-metadata` the original error is suppressed.

If nobody know what's causing this, you could use `gcpMetadata.instance()` instead of `gcpMetadata.isAvailable()`, which would give you access to the error object. 

---

_More info:_

I'm struggling to properly track the source of the problem due to me using yarn workspaces in a multi-module project, my functions are just one module, so when it gets uploaded to GCP it's running `npm install` without a lock file. 

I think it's using 1.3.1 but I'm struggling to prove it. I've spun up a Cloud Shell and downloaded the storage object for my cloud function and ran `npm install` and I don't get `google-auth-library` in the same path as shown in the stacktrace. Instead of it being nested as `@google-cloud/firestore -> google-gax -> google-auth-library`, it's being nested as `google-gax -> google-auth-library`, so although it's likely my function is using 1.3.1, I'm not 100% certain.

**Comment by JustinBeckwith** [comment]
Well this is a bit of a head scratcher :)  First off - thank you filing an excellent bug.  Tons of details, a stack trace, this has it all.  I appreciate that.  

I'm curious - when you say "sporadic" - what does that mean exactly?  Do they happen on an interval, like once an hour?  Or daily?  Or do they come in clusters?  This will make it a lot easier for me to figure out what's going ons.

**Comment by dan-j** [comment]
I'm struggling to see a pattern in when this occurs. It's about 50/50, but may successfully execute 1, 2 or 3 times before failing 1, 2 or 3 times, it just seems random. There isn't very much load on the functions, we're still in development, our functions have anywhere from 1-20 invocations an hour. With none outside business hours.

Digging into Log Viewer it seems that the function is being invoked multiple times. It would appear that the first invocation is successful however I never see a log message I'm expecting via the `@google-cloud/logging` API.

![image](https://user-images.githubusercontent.com/5727701/37497694-13a1a6c4-290e-11e8-9725-d52a326ca5a6.png)

This is what I'm meant to see in the Logs (with my custom log message) when everything is OK, notice the JSON log which has come via the `@google-cloud/logging` API.

![image](https://user-images.githubusercontent.com/5727701/37497718-49fef6a4-290e-11e8-92f1-b1bda1f7b0af.png)

**Comment by dan-j** [comment]
Additionally, I have 8 functions, 2 of which are erroring, another is called frequently which uses the `@google-cloud/logging` library and never errors, and 5 more which rarely get called.

According the "Error Reporting" tool the first occurrence of this was 5th March 11:22 however I can't find a corresponding log at this time. 

The first log I can see with this error is 5th March 16:16 and that version of the function was deployed 14:56 the same day. The deployment was for a different part of the application and none of the actual function code for the 2 misbehaving functions changed, it was just a new commit on master which deploys everything.

Maybe a library somewhere in the dependency chain got updated? Or something changed in the Cloud Function environments around this time?

**Comment by dan-j** [comment]
I've had a thought...

I'm actually using `@google-cloud/logging-bunyan` rather than `@google-cloud/logging` directly, and the difference between the 2 failing functions and the working function is the failing function just logs diffs between a `before` and `after` Firestore snapshot, whereas the working function is doing a fair bit of computation including other database reads/writes. 

My failing functions are creating a logger, log, then return a resolved promise. `logging-bunyan` doesn't give you access to whether the log has been written yet so there's not much else I can do. I think the cloud function may be suspending execution but the logger hasn't finished doing it's work. 

This is where my knowledge of Cloud Functions stops; what would happen to the pending log entry if the function was given 0% CPU? Retrieving the credentials from `admin.credential.applicationDefault()` is synchronous, so will have been loaded before my function returns a promise. If this is indeed the issue, I would have thought I'd have gotten a different error and not "Could not load the default credentials".

Thoughts?

_This is one of the two failing functions, they're almost identical just on different document paths_

```typescript
export const updateUser: CloudFunction<DeltaDocumentSnapshot> = firestore
    .document('/users/{userId}')
    .onUpdate((event) => {
        // abstraction over how to instantiate a logger dependent on environment variables
        // this returns a Logger from @google-cloud/logging-bunyan
        const logger = getRootLogger();

        // log some stuff
        logger.info({ docId: event.data.id }, `updateUser: ${event.data.id}`);

        // return a resolved promise
        return Promise.resolve();
    });
```

**Comment by jkoelker** [comment]
@dan-j I think that is the case as well, chalk it up to my naivety in javascript. I thought `runTransaction` was synchronous, but alas it is not. Thanks for the point in the right direction.

**Comment by JustinBeckwith** [comment]
FYI @ofrobots @google/google-node-team

**Comment by jkoelker** [comment]
After digging for a while, it appears that the documentation for `runTransaction` is incorrect and it is synchronous and not a [`Promise`](https://github.com/googleapis/nodejs-spanner/issues/170) . After adding in more logging, the credential loading error is being passed in as the `err` to my `runTransaction` callback:

```javascript
function objectDelete(file) {                                               
  const database = getDatabase();                                           
  const bucket = file.bucket;                                               
  const size = parseInt(file.size) * -1;                                    
                                                                            
  database.runTransaction((err, transaction) => {                           
    if (err) {                                                              
      console.log('Error starting transaction: ' + err);                    
      return;                                                               
    }                                                                       
                                                                            
    updateBucket(transaction, bucket, size, -1).then(() => {                
      return transaction.commit();                                          
                                                                            
    }).catch(err => {                                                       
      console.log(`Unhandled error in objectFinalize transaction: ${err}`); 
                                                                            
    }).then(() => {
      return database.close();                                              
    });                                                                     
  });                                                                       
}                                                                           
```

The output from the cloudfunc:

```json
{
 "insertId":  "<redacted>" ,
 "labels": {},
 "logName":  "<redacted>" ,
 "receiveTimestamp":  "2018-03-30T19:33:13.751814966Z" ,
 "resource": {},  
 "severity":  "INFO", 
 "textPayload":  "Error starting transaction: Error: Unexpected error while acquiring application default credentials: Could not load the default credentials. Browse to https://developers.google.com/accounts/docs/application-default-credentials for more information." ,
 "timestamp:"  "2018-03-30T19:33:06.851Z" 
}
```

I'm not sure if this is bug is then from this library or from the spanner library, but it is happening prior to my callback for `runTransaction` ever being invoked, and I know its not that its exiting early, since then my callback that does the logging would have never been invoked. I suspect it is in the spanner library as it appears to be using promises extensively. I've brought this up in https://github.com/googleapis/nodejs-spanner/issues/170 there.

**Comment by dan-j** [comment]
@jkoelker that looks asynchronous to me.

just add a return statement to the `database.runTransaction(...` line.

Also, if the Spanner SDK is similar to Firestore (it looks similar, but I haven't used Spanner), the function inside `runTransaction` should return a promise, so add a return statement to your `updateBucket(...` line too.

**Comment by jkoelker** [comment]
@dan-j Unfortunately contrary to the documentation, `database.runTransaction` is not asynchronous and does not return a [promise](https://github.com/googleapis/nodejs-spanner/blob/master/src/database.js#L1263-L1287). It does fire off a series of promises but then all reference are lost by not returning them. I think the cause of this issue is defiantly promises not being completed by the time the cloud func calls `callback()` which terminates the process, causing all pending promises to be rejected.

I think now this is defiantly not an issue with the auth-library, but with the spanner library, perhaps the documentation at https://developers.google.com/accounts/docs/application-default-credentials the error message references could be updated to clarify that this error on cloud funcs indicates just the func is terminating prematurely.

**Comment by erwin16** [comment]
Hi Dan ...  How can you use default credentials as the path to this json file should be set in the GOOGLE_APPLICATION_CREDENTIALS environment variable... and firebase functions:config:set does not accept uppercases ??  thanks for a tip ...

**Comment by dan-j** [comment]
@erwin16 You shouldn't need to set this environment variable for firebase functions because they execute in a sandbox and a call to `admin.credential.applicationDefault()` will use GCP's metadata services to load credentials. 

In any case, you could just write `process.env.GOOGLE_APPLICATION_CREDENTIALS = process.env.google_application_credentials` before calling `.applicationDefault()`

**Comment by paujur** [comment]
Was a solution to this ever found? I can't seem to find one from the comments. I'm getting the same error after deploying functions. It happens the first time I try to call the function after deployment. The second time I try running it, the error is gone, but the function seems to revert to the previous version, and not the version I just deployed (even though it deploys successfully).

Any ideas?

**Comment by JustinBeckwith** [comment]
Greetings folks!  Given the fact that this hasn't seen traffic for a bit, and it looks like the issues are worked out - I am going to go ahead and close this out.  Please do open a new issue if you run into trouble!
